File: apps/sdk-typescript/src/services/negotiation.ts
/**
 * `NegotiationService` — typed client for the `negotiation` service.
 *
 * Mirror of `apps/negotiation/src/http/routes.ts`:
 *   - POST /v1/negotiation                     (start)
 *   - GET  /v1/negotiation/:id                 (get)
 *   - GET  /v1/negotiation/:id/history         (history)
 *   - POST /v1/negotiation/:id/propose         (propose)
 *   - POST /v1/negotiation/:id/counter         (counter)
 *   - POST /v1/negotiation/:id/settle          (settle)
 *
 * `start` accepts an optional `{ idempotencyKey }` second arg per the brief.
 * The key is forwarded into the body verbatim — generation is the caller's
 * responsibility (the service treats `start` as idempotent on this key).
 */

import { request } from '../http.js';

import type { HttpClientOptions } from '../http.js';
import type { IdempotentOptions } from '../types.js';

export type Strategy = 'ascending-auction' | 'multi-criteria';

export type AttributeValue = string | number | boolean | readonly (string | number | boolean)[];

export interface CriterionWeight {
  readonly name: string;
  readonly weight: number;
}

export interface NegotiationTerms {
  readonly subject: string;
  readonly strategy: Strategy;
  readonly constraints?: Readonly<Record<string, AttributeValue>>;
  readonly criteria?: readonly CriterionWeight[];
  readonly partyDids: readonly string[];
  readonly deadline: string;
  readonly reservePrice?: number;
  readonly currency?: string;
}

export interface ProposalInput {
  readonly proposalId: string;
  readonly fromDid: string;
  readonly amount?: number;
  readonly scores?: Readonly<Record<string, number>>;
  readonly payload?: Readonly<Record<string, AttributeValue>>;
  /** base64-encoded Ed25519 signature over the JCS canonical proposal payload. */
  readonly signature: string;
  readonly proposedAt: string;
}

export interface StartRequest {
  readonly terms: NegotiationTerms;
  readonly createdBy: string;
}

export interface ProposeRequest {
  readonly negotiationId: string;
  readonly proposal: ProposalInput;
  /** base64-encoded Ed25519 public key of `proposal.fromDid`. */
  readonly publicKey: string;
}

export interface CounterRequest extends ProposeRequest {
  readonly counterTo: string;
}

export interface SettleSignature {
  readonly did: string;
  readonly signature: string;
}

export interface SettlePublicKey {
  readonly did: string;
  readonly publicKey: string;
}

export interface SettleRequest {
  readonly negotiationId: string;
  readonly winningProposalId?: string;
  readonly signatures: readonly SettleSignature[];
  readonly publicKeys: readonly SettlePublicKey[];
}

export interface NegotiationView {
  readonly negotiationId: string;
  readonly status: string;
  readonly strategy: string;
  readonly terms: {
    readonly subject: string;
    readonly strategy: string;
    readonly constraints: Readonly<Record<string, AttributeValue>>;
    readonly criteria?: readonly CriterionWeight[];
    readonly partyDids: readonly string[];
    readonly deadline: string;
    readonly reservePrice?: number;
    readonly currency?: string;
  };
  readonly partyDids: readonly string[];
  readonly proposals: readonly {
    readonly proposalId: string;
    readonly fromDid: string;
    readonly signature: string;
    readonly proposedAt: string;
    readonly amount?: number;
    readonly scores?: Readonly<Record<string, number>>;
    readonly payload?: Readonly<Record<string, AttributeValue>>;
  }[];
  readonly currentBestProposalId?: string;
  readonly winningProposalId?: string;
  readonly settlementSignatures?: readonly { readonly did: string; readonly signature: string }[];
  readonly createdAt: string;
  readonly updatedAt: string;
  readonly expiresAt: string;
}

export interface HistoryRequest {
  readonly negotiationId: string;
  readonly cursor?: number;
  readonly limit?: number;
}

export interface HistoryView {
  readonly events: readonly {
    readonly seq: number;
    readonly event: Readonly<Record<string, unknown>>;
  }[];
  readonly nextCursor: number | null;
}

export class NegotiationService {
  constructor(
    private readonly opts: HttpClientOptions,
    private readonly baseUrl: string,
  ) {}

  /**
   * POST /v1/negotiation — idempotent on `idempotencyKey`.
   *
   * @example
   *   await client.negotiation.start({ terms, createdBy }, { idempotencyKey: 'k-1' })
   */
  public async start(body: StartRequest, options: IdempotentOptions): Promise<NegotiationView> {
    const data = await request<NegotiationView>(this.opts, {
      method: 'POST',
      baseUrl: this.baseUrl,
      path: '/v1/negotiation',
      body: { ...body, idempotencyKey: options.idempotencyKey },
    });
    if (data === undefined) {
      throw new Error('negotiation.start: empty response body');
    }
    return data;
  }

  /** GET /v1/negotiation/:id */
  public async get(negotiationId: string): Promise<NegotiationView> {
    const data = await request<NegotiationView>(this.opts, {
      method: 'GET',
      baseUrl: this.baseUrl,
      path: `/v1/negotiation/${encodeURIComponent(negotiationId)}`,
    });
    if (data === undefined) {
      throw new Error('negotiation.get: empty response body');
    }
    return data;
  }

  /** GET /v1/negotiation/:id/history?cursor=...&limit=... */
  public async history({ negotiationId, cursor, limit }: HistoryRequest): Promise<HistoryView> {
    const data = await request<HistoryView>(this.opts, {
      method: 'GET',
      baseUrl: this.baseUrl,
      path: `/v1/negotiation/${encodeURIComponent(negotiationId)}/history`,
      query: {
        ...(cursor !== undefined ? { cursor } : {}),
        ...(limit !== undefined ? { limit } : {}),
      },
    });
    if (data === undefined) {
      throw new Error('negotiation.history: empty response body');
    }
    return data;
  }

  /** POST /v1/negotiation/:id/propose */
  public async propose({
    negotiationId,
    proposal,
    publicKey,
  }: ProposeRequest): Promise<NegotiationView> {
    const data = await request<NegotiationView>(this.opts, {
      method: 'POST',
      baseUrl: this.baseUrl,
      path: `/v1/negotiation/${encodeURIComponent(negotiationId)}/propose`,
      body: { proposal, publicKey },
    });
    if (data === undefined) {
      throw new Error('negotiation.propose: empty response body');
    }
    return data;
  }

  /** POST /v1/negotiation/:id/counter */
  public async counter({
    negotiationId,
    counterTo,
    proposal,
    publicKey,
  }: CounterRequest): Promise<NegotiationView> {
    const data = await request<NegotiationView>(this.opts, {
      method: 'POST',
      baseUrl: this.baseUrl,
      path: `/v1/negotiation/${encodeURIComponent(negotiationId)}/counter`,
      body: { counterTo, proposal, publicKey },
    });
    if (data === undefined) {
      throw new Error('negotiation.counter: empty response body');
    }
    return data;
  }

  /** POST /v1/negotiation/:id/settle */
  public async settle({
    negotiationId,
    winningProposalId,
    signatures,
    publicKeys,
  }: SettleRequest): Promise<NegotiationView> {
    const data = await request<NegotiationView>(this.opts, {
      method: 'POST',
      baseUrl: this.baseUrl,
      path: `/v1/negotiation/${encodeURIComponent(negotiationId)}/settle`,
      body: {
        ...(winningProposalId !== undefined ? { winningProposalId } : {}),
        signatures,
        publicKeys,
      },
    });
    if (data === undefined) {
      throw new Error('negotiation.settle: empty response body');
    }
    return data;
  }
}

File: apps/sdk-python/src/praxis_sdk/services/negotiation.py
"""``NegotiationService`` — typed client for the ``negotiation`` service.

Mirror of ``apps/sdk-typescript/src/services/negotiation.ts`` and
``apps/negotiation/src/http/routes.ts``:

- ``POST /v1/negotiation``                     (start — idempotent)
- ``GET  /v1/negotiation/:id``                 (get)
- ``GET  /v1/negotiation/:id/history``         (history)
- ``POST /v1/negotiation/:id/propose``         (propose)
- ``POST /v1/negotiation/:id/counter``         (counter)
- ``POST /v1/negotiation/:id/settle``          (settle)
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any
from urllib.parse import quote

from .._http import HttpClientOptions, RequestParams, request
from ._convert import from_wire, to_wire


@dataclass(frozen=True, slots=True)
class NegotiationProposalView:
    proposal_id: str
    from_did: str
    signature: str
    proposed_at: str
    amount: float | None = None
    scores: dict[str, float] | None = None
    payload: dict[str, Any] | None = None


@dataclass(frozen=True, slots=True)
class NegotiationTermsView:
    subject: str
    strategy: str
    party_dids: list[str]
    deadline: str
    constraints: dict[str, Any] = field(default_factory=dict)
    criteria: list[dict[str, Any]] | None = None
    reserve_price: float | None = None
    currency: str | None = None


@dataclass(frozen=True, slots=True)
class SettlementSignature:
    did: str
    signature: str


@dataclass(frozen=True, slots=True)
class NegotiationView:
    negotiation_id: str
    status: str
    strategy: str
    terms: NegotiationTermsView
    party_dids: list[str]
    proposals: list[NegotiationProposalView]
    created_at: str
    updated_at: str
    expires_at: str
    current_best_proposal_id: str | None = None
    winning_proposal_id: str | None = None
    settlement_signatures: list[SettlementSignature] | None = None


@dataclass(frozen=True, slots=True)
class HistoryEvent:
    seq: int
    event: dict[str, Any]


@dataclass(frozen=True, slots=True)
class HistoryView:
    events: list[HistoryEvent] = field(default_factory=list)
    next_cursor: int | None = None


class NegotiationService:
    """Typed client for the ``negotiation`` service."""

    def __init__(self, opts: HttpClientOptions, base_url: str) -> None:
        self._opts = opts
        self._base_url = base_url

    def start(
        self,
        *,
        terms: dict[str, Any],
        created_by: str,
        idempotency_key: str,
    ) -> NegotiationView:
        """``POST /v1/negotiation`` — idempotent on ``idempotency_key``.

        ``idempotency_key`` is required (the service treats start as
        idempotent on this key — generation is the caller's responsibility).
        """
        body_in: dict[str, Any] = {
            "terms": terms,
            "created_by": created_by,
            "idempotency_key": idempotency_key,
        }
        data = request(
            self._opts,
            RequestParams(
                method="POST",
                base_url=self._base_url,
                path="/v1/negotiation",
                body=to_wire(body_in),
            ),
        )
        if data is None:
            raise RuntimeError("negotiation.start: empty response body")
        return from_wire(NegotiationView, data)

    def get(self, negotiation_id: str) -> NegotiationView:
        """``GET /v1/negotiation/:id``."""
        data = request(
            self._opts,
            RequestParams(
                method="GET",
                base_url=self._base_url,
                path=f"/v1/negotiation/{quote(negotiation_id, safe='')}",
            ),
        )
        if data is None:
            raise RuntimeError("negotiation.get: empty response body")
        return from_wire(NegotiationView, data)

    def history(
        self,
        *,
        negotiation_id: str,
        cursor: int | None = None,
        limit: int | None = None,
    ) -> HistoryView:
        """``GET /v1/negotiation/:id/history?cursor=...&limit=...``."""
        query: dict[str, str | int | float | bool | None] = {}
        if cursor is not None:
            query["cursor"] = cursor
        if limit is not None:
            query["limit"] = limit
        data = request(
            self._opts,
            RequestParams(
                method="GET",
                base_url=self._base_url,
                path=f"/v1/negotiation/{quote(negotiation_id, safe='')}/history",
                query=query if query else None,
            ),
        )
        if data is None:
            raise RuntimeError("negotiation.history: empty response body")
        return from_wire(HistoryView, data)

    def propose(
        self,
        *,
        negotiation_id: str,
        proposal: dict[str, Any],
        public_key: str,
    ) -> NegotiationView:
        """``POST /v1/negotiation/:id/propose``."""
        body = to_wire({"proposal": proposal, "public_key": public_key})
        data = request(
            self._opts,
            RequestParams(
                method="POST",
                base_url=self._base_url,
                path=f"/v1/negotiation/{quote(negotiation_id, safe='')}/propose",
                body=body,
            ),
        )
        if data is None:
            raise RuntimeError("negotiation.propose: empty response body")
        return from_wire(NegotiationView, data)

    def counter(
        self,
        *,
        negotiation_id: str,
        counter_to: str,
        proposal: dict[str, Any],
        public_key: str,
    ) -> NegotiationView:
        """``POST /v1/negotiation/:id/counter``."""
        body = to_wire({"counter_to": counter_to, "proposal": proposal, "public_key": public_key})
        data = request(
            self._opts,
            RequestParams(
                method="POST",
                base_url=self._base_url,
                path=f"/v1/negotiation/{quote(negotiation_id, safe='')}/counter",
                body=body,
            ),
        )
        if data is None:
            raise RuntimeError("negotiation.counter: empty response body")
        return from_wire(NegotiationView, data)

    def settle(
        self,
        *,
        negotiation_id: str,
        signatures: list[dict[str, str]],
        public_keys: list[dict[str, str]],
        winning_proposal_id: str | None = None,
    ) -> NegotiationView:
        """``POST /v1/negotiation/:id/settle``."""
        body_in: dict[str, Any] = {"signatures": signatures, "public_keys": public_keys}
        if winning_proposal_id is not None:
            body_in["winning_proposal_id"] = winning_proposal_id
        data = request(
            self._opts,
            RequestParams(
                method="POST",
                base_url=self._base_url,
                path=f"/v1/negotiation/{quote(negotiation_id, safe='')}/settle",
                body=to_wire(body_in),
            ),
        )
        if data is None:
            raise RuntimeError("negotiation.settle: empty response body")
        return from_wire(NegotiationView, data)
